[Opt Overlap] Support optimizing partial set of parameters

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71620
* __->__ #71608
* #71607
* #71606
* #71605
* #71604
* #71603
* #71602

Typical optimizer usage might be something like `optim.SGD(model.parameters(), lr=1e-2)`, but it is common to freeze a portion of the model so that only a part of it is optimized. One way of implementing this is only passing a subset of parameters you'd like optimized:

```
params_to_opt = [list(model.parameters())[0]]
opt = optim.SGD(params_to_opt, lr=1e-2)
```

However `_register_fused_optim` had no way of supporting this for overlapped optimizers. This diff introduces a `optim_params` argument that can be passed into `_register_fused_optimizer`, allowing only those parameters to be modified during the optimization process. Unittests added to ensure this works as expected. 

Differential Revision: [D33696382](https://our.internmc.facebook.com/intern/diff/D33696382/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

rohan-varma committed:
[Opt Overlap] Support optimizing partial set of parameters

Per title

Differential Revision: [D33696382](https://our.internmc.facebook.com/intern/diff/D33696382/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Opt Overlap] Support optimizing partial set of parameters"


Typical optimizer usage might be something like `optim.SGD(model.parameters(), lr=1e-2)`, but it is common to freeze a portion of the model so that only a part of it is optimized. One way of implementing this is only passing a subset of parameters you'd like optimized:

```
params_to_opt = [list(model.parameters())[0]]
opt = optim.SGD(params_to_opt, lr=1e-2)
```

However `_register_fused_optim` had no way of supporting this for overlapped optimizers. This diff introduces a `optim_params` argument that can be passed into `_register_fused_optimizer`, allowing only those parameters to be modified during the optimization process. Unittests added to ensure this works as expected. 

Differential Revision: [D33696382](https://our.internmc.facebook.com/intern/diff/D33696382/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

cbalioglu commented on torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py:
```diff
@@ -12,25 +12,32 @@ class _OptimizerHookState(object):
         return opt_hook_state_inst
 
+    def _set_params_to_optimize(self, params):
+        self.params_to_optimize = params
```
Nit: This can be simplified to:

```
if params is not None:
    self.params_to_optimize = set(params)
```

rohan-varma replied:
This leads to `self.params_to_optimize` not being defined in the case it is None, but we access it on L60, but we can change that to check `hasattr` as well.

cbalioglu commented on torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py:
```diff
@@ -50,6 +57,7 @@ def _hook_then_optimizer(
     .. warning ::
         This API is experimental adn subject to change.
     """
+    _has_set_params = optimizer_state.params_to_optimize is not None
```
Nit: Since this is a local variable, you can name it `has_set_params` (without an underscore).

cbalioglu reviewed (approved):
Just two nit comments.

cbalioglu commented:
I see that there are two failed distributed tests. Do you mind checking them before shipping this?

rohan-varma commented:
Yep its a missing argument in the test, will be resolved in the next update

rohan-varma committed:
Update on "[Opt Overlap] Support optimizing partial set of parameters"


Typical optimizer usage might be something like `optim.SGD(model.parameters(), lr=1e-2)`, but it is common to freeze a portion of the model so that only a part of it is optimized. One way of implementing this is only passing a subset of parameters you'd like optimized:

```
params_to_opt = [list(model.parameters())[0]]
opt = optim.SGD(params_to_opt, lr=1e-2)
```

However `_register_fused_optim` had no way of supporting this for overlapped optimizers. This diff introduces a `optim_params` argument that can be passed into `_register_fused_optimizer`, allowing only those parameters to be modified during the optimization process. Unittests added to ensure this works as expected. 

Differential Revision: [D33696382](https://our.internmc.facebook.com/intern/diff/D33696382/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma committed:
Update on "[Opt Overlap] Support optimizing partial set of parameters"


Typical optimizer usage might be something like `optim.SGD(model.parameters(), lr=1e-2)`, but it is common to freeze a portion of the model so that only a part of it is optimized. One way of implementing this is only passing a subset of parameters you'd like optimized:

```
params_to_opt = [list(model.parameters())[0]]
opt = optim.SGD(params_to_opt, lr=1e-2)
```

However `_register_fused_optim` had no way of supporting this for overlapped optimizers. This diff introduces a `optim_params` argument that can be passed into `_register_fused_optimizer`, allowing only those parameters to be modified during the optimization process. Unittests added to ensure this works as expected. 

Differential Revision: [D33696382](https://our.internmc.facebook.com/intern/diff/D33696382/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma commented:
CI errors are unrelated - 
"09:41:02 Unable to get pull request builder trigger!!" and "09:10:50 FAIL [0.018s]: test_data_parallel_module_cuda_float16 (__main__.TestDataParallelDeviceTypeCUDA)".
